inspire/dojson/hep/receivers.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2014, 2015 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from inspire.dojson import utils as inspire_dojson_utils
+
+from invenio_records.signals import (
+    before_record_insert,
+    before_record_update,
+)
+
+
+@before_record_insert.connect
+@before_record_update.connect
+def earliest_date(sender, *args, **kwargs):
+    """ Find and assign the earliest date to a HEP paper. """
+    dates = []
+
+    if 'preprint_info' in sender:
+        for preprint_info_key in sender['preprint_info']:
+            if 'date' in preprint_info_key:
+                dates.append(
+                    inspire_dojson_utils.full_date_format(
+                        preprint_info_key['date']))
+
+    if 'defense_date' in sender:
+        for defense_date_key in sender['defense_date']:
+            if 'date' in defense_date_key:
+                dates.append(
+                    inspire_dojson_utils.full_date_format(
+                        defense_date_key['date']))
+
+    if 'thesis' in sender:
+        for thesis_key in sender['thesis']:
+            if 'date' in thesis_key:
+                dates.append(
+                    inspire_dojson_utils.full_date_format(
+                        thesis_key['date']))
+
+    if 'publication_info' in sender:
+        for publication_info_key in sender['publication_info']:
+            if 'year' in publication_info_key:
+                dates.append(
+                    inspire_dojson_utils.full_date_format(
+                        publication_info_key['year']))
+
+    if 'references' in sender:
```

Comment:
I have now tested this branch locally, and I think the logic of getting the year from references is flawed. Simply because it is the date of other papers (and not this one!). E.g. if a paper references an old paper from 1970, then with the current logic the earliest_date is `1970-01-01`. Which clearly is wrong.